openlibrary/solr/update_work.py

```diff
@@ -1433,8 +1442,23 @@ def _solr_update(requests: List[SolrUpdateRequest], commitWithin=60000):
     # Get works for all the editions
     ekeys = set(k for k in keys if k.startswith("/books/"))
 
+    # I wonder if sorting the ekeys could help since it would lead to
+    # requests all being near each other in DB such as 1,2,3,4 etc instead of random
+    # ekeys = list(ekeys)
+    # ekeys.sort()
+
+    timer_results += f"before preload_documents {timer() - start}\n"
+    start = timer()
     data_provider.preload_documents(ekeys)
-    for k in ekeys:
+
+    timer_results += f"preload_documents {timer() - start}\n"
+    start = timer()
+
+    def handle_ekey(k):
+        output_dict = {
+            'deletes': [],
+            'wkeys': set()
+        }
         logger.debug("processing edition %s", k)
         edition = data_provider.get_document(k)
```

Comment:
It is working for me. 
I think this is a bit confusing for a few reasons. Even though it uses the `Parallel` function we are using it in threaded mode so it isn't truly parallel (using different cores) it is concurrent (using different threads). 
If we were using different cores then things would be different and I think you'd need to explicitly pass in the class and it will be cloned onto the new process. However, since we're using threads it will share objects in memory.

In this specific case, since the data_provider is a read only class (as in it only gets data it never changes data elsewhere) I think this is much less likely to have an error. Not sure if that applies to non-default providers.

So 2 questions:
1. Is there anyone in the community that is more familiar with python threading that may be able to touch on this? Maybe @cclauss?
2. Can we do a test with the non-default data_providers?

Thanks!

Reply:
Oh drats, that might be a problem; I just remembered that data_providers also do a lot of caching, and maintaining that cache is important for our prod solr flows. Since we need this to be accessible between threads, we might need to convert some of the `Dict`s the various data_providers use for storage to `ThreadedDict`? This might need a bit more investigation.